include/deal.II/base/lazy.h

```diff
@@ -0,0 +1,346 @@
+   * object.
+   */
+  void
+  reset() noexcept
+  {
+    reset_once_flag();
+    this->std::optional<T>::reset();
+  }
+
+
+#ifndef DOXYGEN
+  /**
+   * Disallow direct emplace()
+   */
+  template <class... Args>
+  T &
+  emplace(Args &&...) = delete;
+
+  /**
+   * Disallow direct emplace()
+   */
+  template <class U, class... Args>
+  T &
+  emplace(std::initializer_list<U>, Args &&...args) = delete;
+#endif
+
+
+  /**
+   * Initialize the wrapped object.
+   *
+   * If the contained object is already initialized this function simply
+   * returns and does nothing.
+   *
+   * If, instead, the object has not yet been initialized then the @p
+   * creator lambda function is called to initialize the contained object.
+   *
+   * This operation is thread safe: The ensure_initialized() method
+   * guarantees that the creator function object is only called once on one
+   * of the calling threads and that after completion the initialization
+   * result (which is stored in the std::optional) is visible on all
+   * threads.
+   */
+  template <typename Callable>
+  DEAL_II_ALWAYS_INLINE inline void
+  ensure_initialized(const Callable &creator)
+  {
+    if (!this->has_value())
+#ifdef DEAL_II_HAVE_CXX20
+      [[unlikely]]
+#endif
+      {
+        // We use std::call_once here to ensure that we call creator()
+        // exactly once on one thread and that the result (creation of
+        // object) is visible on all threads.
+        std::call_once(flag, [&]() {
+          static_cast<std::optional<T> &>(*this) = std::move(creator());
+        });
+      }
+
+    Assert(
+      this->has_value(),
+      dealii::ExcMessage(
+        "The wrapped std::optional<T> object does not contain a valid object "
+        "even though we have just used a call to std::call_once to initialize "
+        "it. Did you accidentally manipulate the underlying std::optional<T> "
+        "object without using the appropriate Lazy<T> interface?"));
+  }
+
+
+  /**
+   * If the std::optional<T> is initialized and contains a valid object the
+   * function simply returns a reference to the contained value. Otherwise,
+   * the @p creator() function object is called to initialize the object
+   * contained in the std::optional<T> first.
+   *
+   * This function mimics the syntax of the std::optional<T> interface and
+   * is functionally equivalent to calling ensure_initialized() followed by
+   * value().
+   *
+   * @note The function is semantically different from the
+   * std::optional<T>::value_or() method as it potentially modifies the
+   * std::optional<T> object by initializing the contained value if needed.
+   *
+   * @post The underlying std::optional<T> is initialized, meaning,
+   * std::optional<T>::has_value() returns true.
+   */
+  template <typename Callable>
+  DEAL_II_ALWAYS_INLINE inline T &
+  value_or_initialize(const Callable &creator)
+  {
+    ensure_initialized(creator);
+    return this->value();
+  }
+
+
+#ifndef DOXYGEN
+private:
```

Comment:
Why do you need this?

Reply:
It was a stylistic choice that I am happy to remove.